### PR TITLE
Minimal staff distances

### PR DESCRIPTION
**Pull request by DavidBauer1984**
Support for minimal staff distances for dynamic and harmonic indications. Distances are measured from the boundary of the staff to the font baseline, see picture below. Works with staffDef and scoreDef attributes as well as command line options.
staffDistances

MEI-example
```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="4.0.0">
   <meiHead>
      <fileDesc>
         <titleStmt>
            <title>Dynamic example</title>
         </titleStmt>
         <pubStmt>
            <date>2017-05-04</date>
         </pubStmt>
      </fileDesc>
   </meiHead>
   <music>
      <body>
         <mdiv>
            <score>
               <scoreDef dynam.dist="6">
                  <staffGrp>
                     <staffDef n="1" lines="5" clef.shape="G" clef.line="2" meter.count="4" meter.unit="4" harm.dist="4" text.dist="24" />
                     <staffDef n="2" lines="5" clef.shape="G" clef.line="2" meter.count="4" meter.unit="4" dynam.dist="12" />
                  </staffGrp>
               </scoreDef>
               <section>
                  <measure n="1">
                     <staff n="1">
                        <layer n="1">
                           <note xml:id="n1" dur="4" oct="4" pname="c" />
                           <note xml:id="n2" dur="4" oct="4" pname="d" />
                           <note xml:id="n3" dur="4" oct="4" pname="e" />
                           <note xml:id="n4" dur="4" oct="4" pname="f" />
                        </layer>
                     </staff>
                     <staff n="2">
                        <layer n="1">
                           <note dur="4" oct="4" pname="c" />
                           <note dur="4" oct="4" pname="d" />
                           <note dur="4" oct="4" pname="e" />
                           <note dur="4" oct="4" pname="f" />
                        </layer>
                     </staff>
                     <dynam staff="1" startid="#n1">p</dynam>
                     <dynam staff="1" startid="#n2">f</dynam>
                     <dynam staff="1" startid="#n3">p</dynam>
                     <dynam staff="1" startid="#n4">f</dynam>
                     <harm staff="1" tstamp="1">C</harm>
                     <dynam staff="2" tstamp="1">p</dynam>
                     <dynam staff="2" tstamp="3">f</dynam>
                  </measure>
                  <measure n="2">
                     <staff n="1">
                        <layer n="1">
                           <note xml:id="n5" dur="4" oct="4" pname="g" />
                           <note xml:id="n6" dur="4" oct="3" pname="g" />
                           <note xml:id="n7" dur="4" oct="4" pname="g" />
                           <note xml:id="n8" dur="4" oct="5" pname="g" />
                        </layer>
                     </staff>
                     <staff n="2">
                        <layer n="1">
                           <note dur="4" oct="4" pname="c" />
                           <note dur="4" oct="4" pname="d" />
                           <note dur="4" oct="4" pname="e" />
                           <note dur="4" oct="4" pname="f" />
                        </layer>
                     </staff>
                     <dynam staff="1" startid="#n5">mf</dynam>
                     <dynam staff="1" startid="#n6">mp</dynam>
                     <dynam staff="1" startid="#n7">fp</dynam>
                     <dynam staff="1" startid="#n8">sfz</dynam>
                     <dynam staff="2" tstamp="1">p</dynam>
                     <dynam staff="2" tstamp="3">f</dynam>
                     <harm staff="1" tstamp="1">G</harm>
                     <harm staff="1" tstamp="3">Gm</harm>
                  </measure>
               </section>
            </score>
         </mdiv>
      </body>
   </music>
</mei>
```

### CODE REVIEW

File: src/doc.cpp [hunk 1]
```diff
@@ -1651,31 +1651,57 @@ double Doc::GetRightMargin(const ClassId classId) const
...
+            }
+        }
+        
+        // #4 Apply CLI option
+        if (classId == DYNAM && m_options->m_dynamDist.IsSet()) {
+            distance = m_options->m_dynamDist.GetValue();
+        }
+        else if (classId == HARM && m_options->m_harmDist.IsSet()) {
+            distance = m_options->m_harmDist.GetValue();
+        }
```
Comment (lpugin) [inline src/doc.cpp:1702]: The code would be more compact if you group them by classId - which you would check only once

File: src/options.cpp [hunk 2]
```diff
@@ -972,7 +972,7 @@ Options::Options()
     this->Register(&m_bracketThickness, "bracketThickness", &m_generalLayout);
 
     m_dynamDist.SetInfo("Dynam dist", "The default distance from the staff for dynamic marks");
-    m_dynamDist.Init(1.0, 0.5, 8.0);
+    m_dynamDist.Init(4.0, 1.0, 16.0);
```
Comment (lpugin) [inline src/options.cpp:975]: Don't change the default without any documented reason

**Review by lpugin — CHANGES_REQUESTED** [review]
Thanks. Next time do make sure to open an issue before making a PR so it can be discussed beforehand.

File: src/options.cpp (hunk 2 above)
Comment (rettinghaus) [reply src/options.cpp:975]: I guess before it was just applied to the top/bottom margin.

Comment (lpugin) [reply src/options.cpp:975]: Maybe. But I can see this, which is not statisfactory

image

Comment (rettinghaus) [reply src/options.cpp:975]: What is actually not satisfactory here is my encoding, as the dim. should be connected to the upper staff. If this is corrected the outcome is this: 
layer_012

Comment (rettinghaus) [reply src/options.cpp:975]: But keeping the old default is also ok.

**Comment by DavidBauer1984** [comment]
I have changed the default of both distances back to 1.0 with boundaries 0.5 and 16.0. Compared to the previous values only the upper boundary is now affected for dynam-dist, which hopefully makes sense, since we need a little more space for distances than for margins.

**Pull request merged**